service/worker/deployment/deployment_util.go

```diff
@@ -0,0 +1,164 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package deployment
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/serviceerror"
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	updatepb "go.temporal.io/api/update/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	deployspb "go.temporal.io/server/api/deployment/v1"
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/payload"
+	"go.temporal.io/server/common/primitives"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/server/common/sdk"
+	"go.temporal.io/server/common/searchattribute"
+)
+
+// GenerateDeploymentWorkflowID is a helper that generates a system accepted
+// workflowID which are used in our deployment workflows
+func GenerateDeploymentWorkflowID(deploymentName string, buildID string) string {
+	return DeploymentWorkflowIDPrefix + DeploymentWorkflowIDDelimeter + deploymentName + DeploymentWorkflowIDDelimeter + buildID
+}
+
+// ValidateDeploymentWfParams is a helper that verifies if the fields used for generating
+// deployment related workflowID's are valid
+func ValidateDeploymentWfParams(fieldName string, field string, maxIDLengthLimit int) error {
+	// Length checks
+	if field == "" {
+		return serviceerror.NewInvalidArgument(fmt.Sprintf("%v cannot be empty", fieldName))
+	}
+
+	// Length of each field should be: (MaxIDLengthLimit - prefix and delimeter length) / 2
+	if len(field) > (maxIDLengthLimit-DeploymentWorkflowIDInitialSize)/2 {
+		return serviceerror.NewInvalidArgument(fmt.Sprintf("size of %v larger than the maximum allowed", fieldName))
+	}
+
+	// Invalid character check - Cannot contain the reserved delimeter ("|") since its being used for building deployment workflowID
+	if strings.Contains(field, DeploymentWorkflowIDDelimeter) {
+		return serviceerror.NewInvalidArgument(fmt.Sprintf("%v cannot contain reserved prefix %v", fieldName, DeploymentWorkflowIDDelimeter))
+	}
+
+	// UTF-8 check
+	return common.ValidateUTF8String(fieldName, field)
+}
+
+// StartAndUpdateDeploymentWorkflow is a helper that validates the input and update workflow parameters
+// and uses a multi-operation to start and update a deployment workflow.
+func StartAndUpdateDeploymentWorkflow(
+	ctx context.Context,
+	workflowArgs *deployspb.DeploymentWorkflowArgs,
+	updateArgs *deployspb.RegisterWorkerInDeploymentArgs,
+	maxIDLengthLimit int,
+	historyClient resource.HistoryClient,
+) (*historyservice.ExecuteMultiOperationResponse, error) {
```

Comment:
It's better to abstract away all the workflow stuff out of physicalTaskQueueManager. That's just an implementation detail to matching. 

I think the right interface to this method should be something like this:
```suggestion

type DeploymentClient interface {
	RegisterWorker(
		ctx context.Context,
		taskQueueName string,
		taskQueueType enumspb.TaskQueueType,
		pollTimestamp time.Time,
	) error
}

type DeploymentWorkflowClient struct {
	namespace  *namespace.Namespace
	deployment *commonpb.WorkerDeployment
	historyClient resource.HistoryClient
}

func NewDeploymentWorkflowClient(
	namespace *namespace.Namespace,
	deployment *commonpb.WorkerDeployment,
	historyClient resource.HistoryClient,
) *DeploymentWorkflowClient {
	return &DeploymentWorkflowClient{...}
}

func (* DeploymentWorkflowClient) RegisterWorker(
	ctx context.Context,
	taskQueueName string,
	taskQueueType enumspb.TaskQueueType,
	pollTimestamp time.Time,
) error {
	// do update with start and all the business. return reasonable error message if something
	// does not go well so the caller can handle the situation accordingly.
}
```

Now physicalTaskQueueManager just needs to create a client via NewDeploymentWorkflowClient and call its method without ever involving itself with the wf business.

Also, later we might decide to move the DeploymentClient stuff in somewhere that is accessible by frontend as well.

Reply:
I agree, thanks for taking the time to jot this down. Helps a lot.